packages/react-devtools-extensions/src/parseHookNames.js

```diff
@@ -0,0 +1,501 @@
+        column: columnNumber,
+      }).line;
+    }
+
+    if (__DEBUG__) {
+      console.log(
+        'findHookNames() mapped line number',
+        lineNumber,
+        'to',
+        originalSourceLineNumber,
+      );
+    }
+
+    if (originalSourceLineNumber === null) {
+      return null;
+    }
+
+    const name = getHookName(
+      hook,
+      hookSourceData.originalSourceAST,
+      ((hookSourceData.originalSourceCode: any): string),
+      ((originalSourceLineNumber: any): number),
+    );
+
+    if (__DEBUG__) {
+      console.log('findHookNames() Found name "' + (name || '-') + '"');
+    }
+
+    map.set(hook, name);
+  });
+
+  return map;
+}
+
+function isValidUrl(possibleURL: string): boolean {
+  try {
+    // eslint-disable-next-line no-new
+    new URL(possibleURL);
+  } catch (_) {
+    return false;
+  }
+  return true;
+}
+
+function loadSourceFiles(
+  fileNameToHookSourceData: Map<string, HookSourceData>,
+): Promise<*> {
+  const promises = [];
+  fileNameToHookSourceData.forEach((hookSourceData, fileName) => {
+    promises.push(
+      fetchFile(fileName).then(runtimeSourceCode => {
+        if (runtimeSourceCode.length > MAX_SOURCE_LENGTH) {
+          throw Error('Source code too large to parse');
+        }
+
+        if (__DEBUG__) {
+          console.groupCollapsed(
+            'loadSourceFiles() fileName "' + fileName + '"',
+          );
+          console.log(runtimeSourceCode);
+          console.groupEnd();
+        }
+
+        hookSourceData.runtimeSourceCode = runtimeSourceCode;
+      }),
+    );
+  });
+  return Promise.all(promises);
+}
+
+async function parseSourceAST(
+  fileNameToHookSourceData: Map<string, HookSourceData>,
+): Promise<*> {
+  // SourceMapConsumer.initialize() does nothing when running in Node (aka Jest)
+  // because the wasm file is automatically read from the file system
+  // so we can avoid triggering a warning message about this.
+  if (!__TEST__) {
+    if (__DEBUG__) {
+      console.log('parseSourceAST() Initializing source-map library ...');
+    }
+
+    // $FlowFixMe
+    const wasmMappingsURL = chrome.extension.getURL('mappings.wasm');
+
+    SourceMapConsumer.initialize({'lib/mappings.wasm': wasmMappingsURL});
+  }
+
+  const promises = [];
+  fileNameToHookSourceData.forEach(hookSourceData => {
+    if (hookSourceData.originalSourceAST !== null) {
+      // Use cached metadata.
+      return;
+    }
+
+    const {runtimeSourceCode, sourceMapContents} = hookSourceData;
+    if (sourceMapContents !== null) {
+      // Parse and extract the AST from the source map.
+      promises.push(
+        SourceMapConsumer.with(
+          sourceMapContents,
+          null,
+          (sourceConsumer: SourceConsumer) => {
+            hookSourceData.sourceConsumer = sourceConsumer;
+
+            // Now that the source map has been loaded,
+            // extract the original source for later.
+            const source = sourceMapContents.sources[0];
```

Comment:
If the compiled code is a bundle, it can consist of multiple modules - what guarantees that `sources[0]` is the correct module for the hook we'll be looking up? The `originalPositionFor` call in `findHookNames` below returns a `source` property which we currently ignore; I think the correct thing to do is to lazily parse/cache the source at that point rather than make this assumption here.

Reply:
Diving into this in #21790, need help debugging my basic test setup though.

Reply:
We also only support one external source map:
https://github.com/facebook/react/blob/ed6c091fe961a3b95e956ebcefe8f152177b1fb7/packages/react-devtools-extensions/src/parseHookNames.js#L224-L230

But maybe that's less of a common concern?

Reply:
I'm not sure what multiple source maps would even mean - a standard source map represents a single, contiguous, compiled file from start to finish. If the browser is executing a file with multiple `sourceMappingURL` comments in it, it's basically a bug in the bundler and should be fixed there. Our best chance is to emulate e.g. Chrome's behaviour if this happens: from a cursory look at V8's [parser](https://github.com/v8/v8/blob/dc712da548c7fb433caed56af9a021d964952728/src/parsing/parser.cc#L3233), I think the *last* one wins.

Reply:
You can see this on Code Sandbox, for example. IIR it generates a single bundle and then injects multiple (inline) source maps.

Reply:
Do you have a repro for multiple `sourceMappingURL`s on CodeSandbox? I did not see multiple `sourceMappingURL` comments in a trivial React sandbox I created just now. I'd love to investigate further and send them a PR if necessary.

Reply:
If I find one again, I'll leave a comment here. I remember seeing one when testing Code Sandbox specifically, but spot checking a few of my recen test apps– I don't see it.